crates/parachain-tokens/src/lib.rs

```diff
@@ -4,120 +4,95 @@
-        type Error = Error<T>;
+        /// The XCM message executor.
+        type XcmExecutor: ExecuteXcm<Self::Call>;
 
-        fn deposit_event() = default;
+        /// The currencies to manage.
+        type MultiCurrency: Into<Vec<u8>> + FullCodec + Clone + Debug + Default + PartialEq;
 
-        /// Transfer collateral to parachain.
-        #[weight = 1000]
-        #[transactional]
-        pub fn transfer_backing_to_parachain(
-            origin,
-            para_id: ParaId,
-            recipient: T::AccountId,
-            network: NetworkId,
-            #[compact] amount: Backing<T>,
-            max_weight: Weight,
-        ) -> DispatchResultWithPostInfo {
-            let who = ensure_signed(origin)?;
+        /// The core balance type.
+        type Balance: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default;
+    }
 
-            if para_id == T::ParaId::get() {
-                return Ok(().into());
-            }
+    // The pallet's events
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    #[pallet::metadata(
+        T::AccountId = "AccountId",
+        T::MultiCurrency = "MultiCurrency",
+        T::Balance = "Balance"
+    )]
+    pub enum Event<T: Config> {
+        /// Transferred currency to parachain.
+        /// [origin, para_id, recipient, network, currency, amount]
+        Transfer(
+            T::AccountId,
+            ParaId,
+            T::AccountId,
+            NetworkId,
+            T::MultiCurrency,
+            T::Balance,
+        ),
+    }
 
-            let raw_amount = Self::tokens_to_u128(amount)?;
-            Self::execute_xcm(
-                AccountId32 {
-                    network: network.clone(),
-                    id: T::AccountId32Convert::convert(who.clone()),
-                }
-                .into(),
-                Self::transfer_to_parachain(
-                    para_id,
-                    recipient.clone(),
-                    network.clone(),
-                    CurrencyId::DOT,
-                    raw_amount
-                ),
-                max_weight,
-            ).map_err(|_| Error::<T>::XcmExecutionFailed)?;
+    #[pallet::error]
+    pub enum Error<T> {
+        XcmExecutionFailed,
+        TryIntoIntError,
+    }
 
-            Self::deposit_event(Event::<T>::TransferBacking(
-                who,
-                para_id,
-                recipient,
-                network,
-                amount,
-            ));
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
 
-            Ok(().into())
-        }
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
 
-        /// Transfer issued tokens to parachain.
-        #[weight = 1000]
+    // The pallet's dispatchable functions.
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Transfer to parachain.
+        #[pallet::weight(1000)]
```

Comment:
We should probably make a benchmark for this, but happy to leave it for a future PR

Reply:
Yep, I'm somewhat inclined to wait for this as we may adopt some pallet from FRAME / cumulus / orml instead.